### PR TITLE
Adding support for `==` and `!=` comparisons for IO channels.

### PR DESCRIPTION
**Pull request by gaurav-singh1998**
Addresses issue #16571 @ben-albrecht @mppf kindly review this whenever you get a chance. Thank you.

### CODE REVIEW

**Review by ben-albrecht — COMMENTED** [review]
Thanks @gaurav-singh1998 - can we add a couple tests for this new behavior?

**Comment by gaurav-singh1998** [comment]
Hi, @ben-albrecht please elaborate the process for writing a test.

**Comment by ben-albrecht** [comment]
@gaurav-singh1998 - see https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst#creating-a-simple-test for the documentation. Following the example of existing tests such as in `test/io` should be helpful too.

**Comment by gaurav-singh1998** [comment]
Hi, @ben-albrecht I have tried adding the appropriate tests kindly review it whenever you have a chance. Thank you.

**Comment by ben-albrecht** [comment]
@gaurav-singh1998 - It's a good practice to double check that Chapel builds and the related tests pass before requesting review.

The CI testing shows that `make check` is failing.

**Comment by gaurav-singh1998** [comment]
Hi, @ben-albrecht sorry for being so late on this. I have rectified the mistakes I made in my previous commits and all the checks are now passing. Kindly review this whenever you have a chance. Thank you.

**Comment by bradcray** [comment]
@gaurav-singh1998 — Following up on our gitter chat the other day, I've manually removed the "cla required" label and we've disabled the script that used to attach the labels, so it shouldn't re-appear (and if it does, help make sure I notice :) ).  Thanks for asking about this.

**Comment by gaurav-singh1998** [comment]
Hi, @bradcray thanks for looking into this matter but it appears that the `chapelautomaton` bot has added the same label again.

**Comment by bradcray** [comment]
Thanks for pointing that out, I'll take another look.

**Comment by bradcray** [comment]
@gaurav-singh1998: Here's a second attempt at removing the label and keeping it removed (where @e-kayrakli did the heavy lifting).

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -1965,6 +1965,30 @@ proc =(ref lhs:channel, rhs:channel) {
...
+  if a.home != b.home then return false;
+  if a._channel_internal != b._channel_internal then return false;
+  if a._readWriteThisFromLocale != b._readWriteThisFromLocale then return false;
+
+  return true;
+}
+
+pragma "no doc"
+proc !=(const ref a: channel, const ref b: channel): bool {
+  if a.writing == b.writing then return false;
```
Comment (mppf) [inline modules/standard/IO.chpl:1982]: Seems to me that this should just return `! (a == b)` to run the above function.

Comment (gaurav-singh1998) [reply modules/standard/IO.chpl]: @mppf I have done the same. Kindly have a look at it. Thanks

Comment (mppf) [reply modules/standard/IO.chpl]: Looks good, thanks

File: test/io/ferguson/channel-comparision.chpl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+use IO;
```
Comment (ben-albrecht) [inline test/io/ferguson/channel-comparision.chpl:1]: typo in filename -- should be:`channel-comparison.chpl` (update `.good` filename too)

File: test/io/ferguson/channel-comparision.chpl [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+proc test1() {
+  var f = opentmp();
+  var a = f.reader();
+  var b:channel(writing=false)=a;
```
Comment (ben-albrecht) [inline test/io/ferguson/channel-comparision.chpl:6]:
```suggestion
  var b:channel(writing=false) = a;
```

File: test/io/ferguson/channel-comparision.good [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Channels a and b are equal.
+Channels a and b are not equal.
```
Comment (ben-albrecht) [inline test/io/ferguson/channel-comparision.good:2]: This `.good` file probably needs a newline (though I haven't confirmed with testing)

**Review by ben-albrecht — CHANGES_REQUESTED** [review]
@gaurav-singh1998 - we are catching up on some reviews, thanks for your patience.

I am seeing the following compile error for your newly added test:

```
[Executing diff channel-comparision.good channel-comparision.comp.out.tmp]
1,2c1,9
< Channels a and b are equal.
< Channels a and b are not equal.
\ No newline at end of file
---
> $CHPL_HOME/modules/standard/IO.chpl:1969: In function '==':
> $CHPL_HOME/modules/standard/IO.chpl:1974: error: unresolved call '!=(qio_channel_ptr_t, qio_channel_ptr_t)'
> $CHPL_HOME/modules/standard/IO.chpl:1981: note: this candidate did not match: !=(const ref a: channel, const ref b: channel)
> $CHPL_HOME/modules/standard/IO.chpl:1974: note: because call actual argument #1 with type qio_channel_ptr_t
> $CHPL_HOME/modules/standard/IO.chpl:1981: note: is passed to formal 'const ref a: channel'
> $CHPL_HOME/modules/standard/IO.chpl:729: note: candidates are: !=(_arg1: qiovec_t, _arg2: qiovec_t)
> $CHPL_HOME/modules/standard/IO.chpl:768: note:                 !=(_arg1: iostyle, _arg2: iostyle)
> note: and 119 other candidates, use --print-all-candidates to see them
>   channel-comparision.chpl:7: called as ==(a: channel(false,dynamic,true), b: channel(false,dynamic,true))
[Error matching compiler output for io/ferguson/channel-comparision]
```

This is because `!=` is not defined for the `qio_channel_ptr` type.

These IO types predate the CPtr stuff, so we should utilize that to get a definition for `!=` on `qio_channel_ptr`. Probably `extern type qio_channel_ptr_t;` etc. should utilize `extern record qio_channel_t` and then make the type`type qio_channel_ptr = c_ptr(qio_channel_struct)`.

See https://chapel-lang.org/docs/master/technotes/extern.html for some background on C interop in Chapel. Let us know if you have any questions. Thanks!

**Comment by ben-albrecht** [comment]
@gaurav-singh1998 - I hope you find time to return to this PR eventually - it would be a nice addition. I am going to close the PR for now due to inactivity to help keep our active PR list manageable. We can reopen when you are able to look at it again.

**Pull request closed without merging**